src/algebra/parity.lean

```diff
@@ -8,50 +8,121 @@ import algebra.ring.basic
 import algebra.algebra.basic
 import algebra.group_power.basic
 import algebra.field_power
+import algebra.opposites
 
-/-!  This file proves some general facts about even and odd elements of semirings. -/
+/-!  # Squares, even and odd elements
 
+This file proves some general facts about squares, even and odd elements of semirings.
+
+In the implementation, we define `is_square` and we let `even` be the notion transported by
+`to_additive`.  The definition are therefore as follows:
+```lean
+is_square a ↔ ∃ r, a = r * r
+even a ↔ ∃ r, a = r + r
+```
+
+Odd elements are not unified with a multiplicative notion.
+
+* TODO: Try to generalize further the typeclass assumptions on `is_square/even`.
+  For instance, in some cases, there are `semiring` assumptions that I (DT) am not convinced are
+  necessary.
+* TODO: Consider moving the definition and lemmas about `odd` to a separate file.
+* TODO: The "old" definition of `even a` asked for the existence of an element `c` such that
+  `a = 2 * c`.  For this reason, several fixes introduce an extra `two_mul` or `← two_mul`.
+  It might be the case that by making a careful choice of `simp` lemma, this can be avoided.
+ -/
+
+open mul_opposite
 variables {α β : Type*}
 
-section semiring
-variables [semiring α] [semiring β] {m n : α}
+/--  An element `a` of a type `α` with multiplication satisfies `square a` if `a = r * r`,
+for some `r : α`. -/
+@[to_additive
+"An element `a` of a type `α` with addition satisfies `even a` if `a = r + r`,
+for some `r : α`."]
+def is_square [has_mul α] (a : α) : Prop := ∃ r, a = r * r
 
-/-- An element `a` of a semiring is even if there exists `k` such `a = 2*k`. -/
-def even (a : α) : Prop := ∃ k, a = 2*k
+@[simp, to_additive]
+lemma is_square_mul_self [has_mul α] (m : α) : is_square (m * m) := ⟨m, rfl⟩
 
-@[simp] lemma even_zero : even (0 : α) := ⟨0, (mul_zero _).symm⟩
+@[to_additive even_iff_exists_two_nsmul]
+lemma is_square_iff_exists_sq [monoid α] (m : α) : is_square m ↔ ∃ c, m = c ^ 2 :=
+by simp [is_square, pow_two]
```

Comment:
It's probably worth adding a dot-notation version of this lemma, something like:
```lean
alias is_square_iff_exists_sq ↔ is_square.exists_sq is_square_of_exists_sq
```

so you can replace usages like
```lean
    rcases nat.even.sub_even hn (even_bit0 1) with ⟨k, hk⟩,
    rw ← two_mul at hk,
```
with
```lean
    rcases (nat.even.sub_even hn (even_bit0 1)).exists_two_mul with ⟨k, hk⟩,
```

Reply:
Thanks!  I got the alias working: I had to explicitly use `attribute [to_additive ...] ...`, since the name was not autogenerated.